### ISSUE TITLE
Search: Fix Facet Filtering - Checking and Unchecking Facet Behavior is Off

### ISSUE THREAD

**Issue by odlbot** [mitodl/open-discussions#2975]
As an Open users, I want to search for resources by selecting facets, so that I can find them quickly. There are some issues with how facets are filtered and unfiltered when they are selected and unselected. See below for Steps to Reproduce and video. 

**Steps to Reproduce**

1. Click on ‘MITx’ in Offered By
1. Click on ‘Videos’ (at top) - Note the options that filter in Offered By
1. Unclick ‘Videos’ - Note that the options don’t all come back in Offered By to where they were before selecting ‘Videos’

See the attached video. 

Note that this behavior happens with many different combinations of facets from in different groups. (e.g. Courses + Open Content)

The basic rule should be when you select a facet, and then uncheck it, it should return to the state it was before checked. 

**Acceptance Criteria:**

* [ ] Reproduce the issue
* [ ] Confirm the desired behavior for filtering facets when boxes are checked
* [ ] Fix this issue

**Out of Scope**

**Dependencies**

┆Attachments: <a href="https://trello-attachments.s3.amazonaws.com/5dfd295df7c63e06791c4cae/5ed5564bd1f6507aa6e6317f/0aa2ef3ce14d0200ad2202f4fa90bc21/OpeFacetsBehavior.mov">OpeFacetsBehavior.mov</a>

**Comment by aliceriot** [comment]
so there are some reasons why it works that way. I will try to explain but @mbertrand may have some input too.

basically the thing with it is that on the front-end we don't have a list of all the possible facet values that we store anywhere. instead, we look at the different values that come back in the search response and present all of them as options.

4:29
so across all the different results we'll have a certain number of results for the 'Offered By' field





4:31
and whatever things are there in the response will be what we can renderwhen you select a facet value, we store all of the options of the most recently selected facet from the time before your last selection





4:33
otherwise there is a weird user experience where, for instance, as soon as you click 'MITx' all of the other options would disappear





4:33
because if you're filtering for 'MITx' then you won't get all of those other things in the search response

**Comment by mbertrand** [comment]
@aliceriot I think you explained it well.  The facets shown are based on what elasticsearch returns based on the current filters, if there are 0 results for a particular option based on current filters then it won't be displayed.  The only exception to that is that when you click on a selection within a particular facet group, for example clicking 'MITx' in the "Offered By' facet group, the other options within that group are not yet removed.  But options in all the other facet groups which have 0 results are removed.  It's a long time ago but I think it was purposely designed to work this way.

**Comment by mbertrand** [comment]
It might be possible to only add new options to facet groups as they are returned from Elasticsearch results, but never remove them, if that's closer to the desired behavior.

**Comment by briangrossman** [comment]
Thank you @aliceriot and @mbertrand   

This mostly makes sense to me, but there's still a little I don't quite understand. Apologies.

Stepping through the example in the video, here's how I think it's working. Please let me know what I'm missing. 

**Step 1: Arrive at search**
Everything is viewable nothing is filtered. Got it.

**Step 2: Click on MITx**
After clicking MITx, if there are any facets in other groups (i.e. not Offered By) that don't have any resources that are aligned with them && MITx (e.g. Social Medicine), they disappear. But, it appears as though all the Offered By facets remain. Got it. 

**Step 3: Click on Videos**
After clicking on Videos (MITx still selected), it only shows facets that have resources that are aligned with both OfferedBy==MITx && Learning Resources==Videos. Ok. This makes sense, too. Got it.

**Step 4: Uncheck Videos**
Now, only MITx remains checked. My expectation would be that this is the same situation as Step 2 when only MITx was selected, but now, there are missing options from OfferedBy that weren't missing the first time only MITx was selected. All that remains is MITx, MicroMasters, and OCW. Why is that? Are those the only facets that have resources that both MITx and (Micromasters or OCW)? Why would having just MITx selected have different facets available depending on how you got there?

**Comment by aliceriot** [comment]
yes-  this is the tricky part.

1) Are those the only facets that have resources that both MITx and (Micromasters or OCW)?

yes this is what's going on here. when you have MITx selected you will only see options in the 'offered by' thing that come back in the result set when you filter for MITx. So, in other words, you'll get MITx, and you'll see Micromasters and OCW because some items have multiple offered_by entries.

2) Why would having just MITx selected have different facets available depending on how you got there?

This is the complicated bit. We have a little bit of a special exception to the "only show facet options that come up in the result set" rule.

Basically, if you just implemented that rule straight up then, in step 2 above all options except Micromasters and OCW should disappear from the 'offered by' options list as soon as you clicked 'MITx', because we update the search right away. If this were the behavior then it wouldn't be possible to select two different values for 'offered by' which did not 'intersect', because after selecting the first one the second one would disappear from the UI.

So we have a little workaround, where we store the facet options for the most recently clicked facet. We then display, in the UI, these stored values as options instead of just displaying the values that come back in the search results. this lets the user select, say, two values for offered by' which do not have any overlap.

but we only do this for the most recently clicked facet, and _not_ for all the facets. so in the flow above what's going on is that, after step 2, we store the options that were present in the initial result set _only_ for the offered by facet, so after clicking MITx the other options are still present.

But then when 'videos', which is an option under a different facet, is selected that wipes out the stored values for the 'offered by' facet, so then when 'videos' in unchecked we don't have the stored values around anymore and we _only_ display the values which come back in the search response.

Does that make sense? It is a weirdly specific behavior that we wanted to have because if we _dont_ have it then it's impossible to select 'video' and 'course' for learning resource type, for instance. but it's a little mind bending! so if this explanation doesn't make sense I can try to break it down differently.

here is the issue for the change when we made it: https://github.com/mitodl/open-discussions/issues/1917
and here's the PR: https://github.com/mitodl/open-discussions/pull/1928

**Comment by abeglova** [comment]
I think the correct behavior for the facets options is for each facet to display the available options after filtering by the selections in other facets but not that facet. So the aggregation query for each facets needs to be different from the query to filter results.

**Comment by abeglova** [comment]
In the above example, clicking on "MITx" in the Offered By facet will affect the available options and document counts in all the facets except for Offered By. That means that the aggregation query parameters need to be different for every facet with at least one item selected

**Comment by aliceriot** [comment]
ok, let me try to make sure I understand the proposal. in the scenario that Brian posted above would that look something like this?

**Step 1: Arrive at search**
We show all values for all the facets.

**Step 2: Click on MITx**
all facets except 'offered by' will change to show _only_ the values which are found in objects matching `offered_by === "MITx"`. all values will continue to be shown in 'offered by'.

**Step 3: Click on Videos**
this is in the 'Learning Resource' facet. Now we'll have in:

- Learning Resources: the list of learning resource types found in the result set of `offered_by === "MITx"`
- Offered By: the list of platforms / authors found in the result set of `object_type === "video"`

both reflect for facet _x_ the result set _R(Y)_ where _Y_ is the set of all facets minus _x_. So for 'Learning resources' the values are based on doing a query solely on `offered_by`, whereas for 'Offered by' the values are based on doing a query solely for `object_type`.

**Step 4: Uncheck Videos**
Now we have only `"MITx"` selected under 'Offered by'. We'll show these options:

- Offered by: all possible options, because we don't have options selected on any other facet.
- all other facets: only the options found in the result set of `offered_by === "MITx"`

two questions I have:

- @abeglova does that sound like it's in line with what you're proposing?
- if I got it right, can we implement this behavior easily with elasticsearch?
- if we can, what would the performance implications look like of doing so? It seems like if we have a value selected in each facet then we'll need to run an additional query for each facet to find the values to show for that one
- for @briangrossman and @Ferdi what do you all think? @mbertrand interested in your feedback too

**Comment by abeglova** [comment]
1) Yes, that is the behavior I am proposing 2) It can definitely be implemented with elasticsearch fairly easily 3) I'm not sure what the performance will be like. I think aggregation queries should be fast enough that it won't be a problem but I haven't tested it

**Comment by briangrossman** [comment]
From a user perspective, I think this solves the original issue. I like it. 

The issue thing I don't love about the current implementation boils down to the  fact that you can be on the page with only MITx selected and have different results depending on how you got there. This solves that.

I agree with your questions though, @aliceriot. Performance is the big question.

**Comment by briangrossman** [comment]
Thinking ahead, in case we get to the point where we are sorting through examples, I threw together a simple dataset we can use for clarity:

### Facets
**Color**
- Red
- Green


**Food Group**
- Fruit
- Vegetable
- Meat


**Needs Refrigeration**
- Yes
- No


### Data

- Apple (Red or Green, Fruit, Doesn't Need Refrigeration)
- Strawberries (Red, Fruit, Needs Refrigeration)
- Cranberries (Red, Fruit, Doesn't Needs Refrigeration)
- Pepper (Red or Green, Vegetable, Needs Refrigeration)
- Avocado (Green, Vegetable, Doesn't Need Refrigeration)
- Lettuce (Green, Vegetable, Needs Refrigeration)
- Steak (Red, Meat, Needs Refrigeration)
- Live Fox (Red, Meat, Doesn't need Refrigeration)
- Frog Legs (Green, Meat, Needs Refrigeration)
- Live Frog (Green, Meat, Doesn't Need Refrigeration)

**Comment by odlbot** [comment]
➤ Brian Grossman commented:

Additional details: https://mitodl.slack.com/archives/G5RHT8GDD/p1591366748056700
Ferdi found a related issue

**Comment by odlbot** [comment]
➤ Brian Grossman commented:

Issue 1

  1. Click Social Medicine (the last Subject)
  2. Click OCW (Offered By)
  3. UnClick Social Medicine - note that the ordering is wrong for subjects (new ones just flow underneath)

This issue happens one any of the facet groups.

Issue 2

  1. Click OCW (Offered By)
  2. Click Engineering (top Subject Area)
  3. Unclick Engineering - note that MITx has 1 match
  4. Unclick OCW - MITx still has 1 match
